[quant][pt2e] Refactor node.meta["target_dtype_info"] to allow generalized observer/fq configuration for input

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97483
* __->__ #97472

Summary:
This is a follow up PR to generalize the way of how users could configure what observer/fq class they want to use for each input of the node,
follow up item (1) in https://github.com/pytorch/pytorch/pull/92641

Previously people will configure the observer/fq class in the following manner:

node.meta["target_dtype_info"] = {
  "input_act_obs_or_fq_ctr": obs1,
  "weight_obs_or_fq_ctr": obs2,
  "bias_obs_or_fq_ctr": obs3,
  "output_obs_or_fq_ctr": obs4,
}

This has hardcoded "weight" and "bias" and we are storing the index for weight and bias in BackendConfig,

this works for existing use cases but it's not general enough, since people may need to observe non-weight Tensors, and it's possible to use different
types of observers for each of them, and the input args may have more complicated structure.
```
node.meta["target_dtype_info"] = {
  "input_obs_or_fq_ctrs": (obs1, obs2, obs3),
  "output_obs_or_fq_ctr": obs4,
}

node.meta["target_dtype_info"] = {
  # this matches node.args
  "input_obs_or_fq_ctrs": (obs1, (obs2, obs3),)
  "output_obs_or_fq_ctr": obs4,
}
```

Next: do similar changes for output

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D44389339](https://our.internmc.facebook.com/intern/diff/D44389339)

jerryzh168 committed:
[quant][pt2e] Refactor node.meta["target_dtype_info"] to allow generalized observer/fq configuration for input

Summary:
This is a follow up PR to generalize the way of how users could configure what observer/fq class they want to use for each input of the node,
follow up item (1) in https://github.com/pytorch/pytorch/pull/92641

Previously people will configure the observer/fq class in the following manner:

node.meta["target_dtype_info"] = {
  "input_act_obs_or_fq_ctr": obs1,
  "weight_obs_or_fq_ctr": obs2,
  "bias_obs_or_fq_ctr": obs3,
  "output_obs_or_fq_ctr": obs4,
}

This has hardcoded "weight" and "bias" and we are storing the index for weight and bias in BackendConfig,

this works for existing use cases but it's not general enough, since people may need to observe non-weight Tensors, and it's possible to use different
types of observers for each of them, and the input args may have more complicated structure.

node.meta["target_dtype_info"] = {
  "input_obs_or_fq_ctrs": (obs1, obs2, obs3),
  "output_obs_or_fq_ctr": obs4,
}

node.meta["target_dtype_info"] = {
  # this matches node.args
  "input_obs_or_fq_ctrs": (obs1, (obs2, obs3),)
  "output_obs_or_fq_ctr": obs4,
}

Next: do similar changes for output

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Refactor node.meta["target_dtype_info"] to allow generalized observer/fq configuration for input"


Summary:
This is a follow up PR to generalize the way of how users could configure what observer/fq class they want to use for each input of the node,
follow up item (1) in https://github.com/pytorch/pytorch/pull/92641

Previously people will configure the observer/fq class in the following manner:

node.meta["target_dtype_info"] = {
  "input_act_obs_or_fq_ctr": obs1,
  "weight_obs_or_fq_ctr": obs2,
  "bias_obs_or_fq_ctr": obs3,
  "output_obs_or_fq_ctr": obs4,
}

This has hardcoded "weight" and "bias" and we are storing the index for weight and bias in BackendConfig,

this works for existing use cases but it's not general enough, since people may need to observe non-weight Tensors, and it's possible to use different
types of observers for each of them, and the input args may have more complicated structure.
```
node.meta["target_dtype_info"] = {
  "input_obs_or_fq_ctrs": (obs1, obs2, obs3),
  "output_obs_or_fq_ctr": obs4,
}

node.meta["target_dtype_info"] = {
  # this matches node.args
  "input_obs_or_fq_ctrs": (obs1, (obs2, obs3),)
  "output_obs_or_fq_ctr": obs4,
}
```

Next: do similar changes for output

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Refactor node.meta["target_dtype_info"] to allow generalized observer/fq configuration for input"


Summary:
This is a follow up PR to generalize the way of how users could configure what observer/fq class they want to use for each input of the node,
follow up item (1) in https://github.com/pytorch/pytorch/pull/92641

Previously people will configure the observer/fq class in the following manner:

node.meta["target_dtype_info"] = {
  "input_act_obs_or_fq_ctr": obs1,
  "weight_obs_or_fq_ctr": obs2,
  "bias_obs_or_fq_ctr": obs3,
  "output_obs_or_fq_ctr": obs4,
}

This has hardcoded "weight" and "bias" and we are storing the index for weight and bias in BackendConfig,

this works for existing use cases but it's not general enough, since people may need to observe non-weight Tensors, and it's possible to use different
types of observers for each of them, and the input args may have more complicated structure.
```
node.meta["target_dtype_info"] = {
  "input_obs_or_fq_ctrs": (obs1, obs2, obs3),
  "output_obs_or_fq_ctr": obs4,
}

node.meta["target_dtype_info"] = {
  # this matches node.args
  "input_obs_or_fq_ctrs": (obs1, (obs2, obs3),)
  "output_obs_or_fq_ctr": obs4,
}
```

Next: do similar changes for output

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -119,12 +119,12 @@
 # should be moved to the new programmable API class soon
 _DEFAULT_FP32_QCONFIG_FOR_TARGET_DTYPE_INFO = {
-    "input_act_obs_or_fq_ctr": torch.ao.quantization.qconfig._default_fp32_placeholder_qconfig.activation,
+    "input_obs_or_fq_ctr_for_args": torch.ao.quantization.qconfig._default_fp32_placeholder_qconfig.activation,
```
Can we drop training `_for_args`? Also what does ctr stand for?

kimishpatel replied:
Also there is not real "default", right? Because now each node can have different input structure. If this is meant for single tensor input qconfig, maybe name it as "default_fp32_tensor_qconfig.."

jerryzh168 replied:
I'm using `_for_args` since we have "_for_kwargs", would it be confusing if we drop it?

jerryzh168 replied:
`ctr` stands for constructor

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -119,12 +119,12 @@
 
 _DEFAULT_QUINT8_QCONFIG_FOR_TARGET_DTYPE_INFO = {
-    "input_act_obs_or_fq_ctr": torch.ao.quantization.qconfig._default_quint8_placeholder_qconfig.activation,
+    "input_obs_or_fq_ctr_for_args": torch.ao.quantization.qconfig._default_quint8_placeholder_qconfig.activation,
```
Same here as prev comment.

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -475,6 +512,43 @@ def _get_arg_target_dtype_as_output(
+) -> None:
+    if not isinstance(input_obs_or_fq_ctr_for_args, tuple):
+        # if input_obs_or_fq_ctr_for_args is a single element,
+        # use the same obs_or_fq_ctr for all args
```
This is fine for now but I dont think we want to provide API where you use single observer that is used for all args. Just be explicit and require each arg's observer be specified. I think that what we agreed for prepare's interface?

jerryzh168 replied:
yes, this is for bc purposes, it is not recommended for new users to use this

kimishpatel replied:
So I think we will need sort of something like:
Old_API -> populate_args_to_ctr this breaks without kwargs"
New_API -> something_else_which_provides_no_way_for_kwargs"
"populate_args_to_ctr this breaks without kwargs"  -> "rest of prepare"
"something else which provides no way for kwards" -> "rest of prepare"

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -475,6 +512,43 @@ def _get_arg_target_dtype_as_output(
+        elif isinstance(arg_or_args, Node):
+            arg_to_ctr[arg_or_args] = input_obs_or_fq_ctr_for_args
+    else:
+        for arg, input_obs_or_fq_ctr in zip(arg_or_args, input_obs_or_fq_ctr_for_args):
```
Also, how we gurantee that coming here means both `arg_or_args` and `input_obs_or_fq_ctr_for_args` have the same structure. I think we should enforce that here such that we ask for the signatures to match.

kimishpatel replied:
I see you address this later

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -475,6 +512,43 @@ def _get_arg_target_dtype_as_output(
+        for arg, input_obs_or_fq_ctr in zip(arg_or_args, input_obs_or_fq_ctr_for_args):
+            if isinstance(arg, tuple):
+                if isinstance(input_obs_or_fq_ctr, tuple):
+                    # TODO: have a separate validation pass for this
```
Ok, thats what I meant in the previous comment. But you already have a todo, so ignore what i said in previous comment.

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -485,24 +559,16 @@ def _get_arg_target_dtype_as_input_to_node(
+    input_obs_or_fq_ctr_for_args = node.meta["target_dtype_info"]["input_obs_or_fq_ctr_for_args"]
+    arg_to_ctr = {}
+    _populate_arg_to_ctr_for_args(node.args, input_obs_or_fq_ctr_for_args, arg_to_ctr)
+    if "_input_obs_or_fq_ctr_for_kwargs" in node.meta["target_dtype_info"]:
```
This will eventually go away for aten graph, right? I am just wondering if we should take a short cut and either do simpler things instead of matching structures, or just assert?

Actually I do see that we are doing simple things, so seems ok to me.

jerryzh168 replied:
yeah this is outdated, I changed it to assert

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -485,24 +559,16 @@ def _get_arg_target_dtype_as_input_to_node(
-        return qconfig_bias_dtype
+    assert "target_dtype_info" in node.meta
+    assert "input_obs_or_fq_ctr_for_args" in node.meta["target_dtype_info"]
+    input_obs_or_fq_ctr_for_args = node.meta["target_dtype_info"]["input_obs_or_fq_ctr_for_args"]
```
```suggestion
    obs_of_fq_config = node.meta["target_dtype_info"]
```

kimishpatel replied:
Might make the subsequent code easier to read

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -159,13 +159,23 @@ def _is_input_arg_dtype_supported_by_backend(
+        _populate_arg_to_ctr_for_kwargs(node.kwargs, node.meta["target_dtype_info"]["_input_obs_or_fq_ctr_for_kwargs"], arg_to_ctr)
+
     # TODO: support check for standalone module
+    # TODO: remove the need to do weight/bias/act specific check after refactoring
```
Cool!

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -159,13 +159,23 @@ def _is_input_arg_dtype_supported_by_backend(
+    assert "input_obs_or_fq_ctr_for_args" in node.meta["target_dtype_info"]
+    input_obs_or_fq_ctr_for_args = node.meta["target_dtype_info"]["input_obs_or_fq_ctr_for_args"]
+    arg_to_ctr = {}
+    _populate_arg_to_ctr_for_args(node.args, input_obs_or_fq_ctr_for_args, arg_to_ctr)
```
I also presume this will be simplified because right now we seem to be doing this check `_is_input_arg_dtype_supported_by_backend` for every arg of a node. Anyways with new API, this should not be needed, right?

jerryzh168 replied:
yeah, some of this are kept for bc purposes, with the new API and improved BackendConfig, we don't need to have separate checkings for act/weight/bias

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -437,10 +447,37 @@ def _get_target_activation_dtype_for_node(
-            "input_act_obs_or_fq_ctr": qconfig.activation,
-            "weight_obs_or_fq_ctr": qconfig.weight,
-            "bias_obs_or_fq_ctr": PlaceholderObserver.with_args(dtype=bias_dtype),
+            "input_obs_or_fq_ctr_for_args": tuple(input_obs_or_fq_ctr_for_args),
```
Does this change actually work for all nodes? I am just wondeirng if this is going to break any existing usecase.

jerryzh168 replied:
yeah I have run all existing tests to make sure this doesn't break anything, will also import to phabricator and test the internal things as well before landing

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -437,10 +447,37 @@ def _get_target_activation_dtype_for_node(
+            "input_obs_or_fq_ctr_for_args": tuple(input_obs_or_fq_ctr_for_args),
+            # this is kept for fx graph mode quantization use cases, not supposed
+            # to be used for quantize_pt2e
+            "_input_obs_or_fq_ctr_for_kwargs": {"weight": qconfig.weight, "bias": BIAS_OBS_OR_FQ_CTR},
```
Same here. Which nodes have such kwargs?

kimishpatel replied:
Ok I see your comment, so it is useful somewhere? Curious where

jerryzh168 replied:
here is the test that breaks without this: (https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py#L7331)

kimishpatel replied:
is it because of weight/bias on layer_norm?

jerryzh168 replied:
> here is the test that breaks without this: (https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py#L7331)

@kimishpatel here is the test the need kwargs config

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -514,15 +580,16 @@ def _get_arg_target_is_dynamic_as_input_to_node(
+    arg_to_ctr = {}
+    _populate_arg_to_ctr_for_args(node.args, input_obs_or_fq_ctr_for_args, arg_to_ctr)
+    if "_input_obs_or_fq_ctr_for_kwargs" in node.meta["target_dtype_info"]:
+        _populate_arg_to_ctr_for_kwargs(node.kwargs, node.meta["target_dtype_info"]["_input_obs_or_fq_ctr_for_kwargs"], arg_to_ctr)
```
As before, we are repeatedly populating the dictionary. If you think this will be used later then consider refactoring in subsequent diff.

jerryzh168 replied:
sounds good, I'll refactor this

kimishpatel reviewed (changes requested):
Left some comments

jerryzh168 committed:
Update on "[quant][pt2e] Refactor node.meta["target_dtype_info"] to allow generalized observer/fq configuration for input"


Summary:
This is a follow up PR to generalize the way of how users could configure what observer/fq class they want to use for each input of the node,
follow up item (1) in https://github.com/pytorch/pytorch/pull/92641

Previously people will configure the observer/fq class in the following manner:

node.meta["target_dtype_info"] = {
  "input_act_obs_or_fq_ctr": obs1,
  "weight_obs_or_fq_ctr": obs2,
  "bias_obs_or_fq_ctr": obs3,
  "output_obs_or_fq_ctr": obs4,
}

This has hardcoded "weight" and "bias" and we are storing the index for weight and bias in BackendConfig,

this works for existing use cases but it's not general enough, since people may need to observe non-weight Tensors, and it's possible to use different
types of observers for each of them, and the input args may have more complicated structure.
```
node.meta["target_dtype_info"] = {
  "input_obs_or_fq_ctrs": (obs1, obs2, obs3),
  "output_obs_or_fq_ctr": obs4,
}

node.meta["target_dtype_info"] = {
  # this matches node.args
  "input_obs_or_fq_ctrs": (obs1, (obs2, obs3),)
  "output_obs_or_fq_ctr": obs4,
}
```

Next: do similar changes for output

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Refactor node.meta["target_dtype_info"] to allow generalized observer/fq configuration for input"


Summary:
This is a follow up PR to generalize the way of how users could configure what observer/fq class they want to use for each input of the node,
follow up item (1) in https://github.com/pytorch/pytorch/pull/92641

Previously people will configure the observer/fq class in the following manner:

node.meta["target_dtype_info"] = {
  "input_act_obs_or_fq_ctr": obs1,
  "weight_obs_or_fq_ctr": obs2,
  "bias_obs_or_fq_ctr": obs3,
  "output_obs_or_fq_ctr": obs4,
}

This has hardcoded "weight" and "bias" and we are storing the index for weight and bias in BackendConfig,

this works for existing use cases but it's not general enough, since people may need to observe non-weight Tensors, and it's possible to use different
types of observers for each of them, and the input args may have more complicated structure.
```
node.meta["target_dtype_info"] = {
  "input_obs_or_fq_ctrs": (obs1, obs2, obs3),
  "output_obs_or_fq_ctr": obs4,
}

node.meta["target_dtype_info"] = {
  # this matches node.args
  "input_obs_or_fq_ctrs": (obs1, (obs2, obs3),)
  "output_obs_or_fq_ctr": obs4,
}
```

Next: do similar changes for output

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Refactor node.meta["target_dtype_info"] to allow generalized observer/fq configuration for input"


Summary:
This is a follow up PR to generalize the way of how users could configure what observer/fq class they want to use for each input of the node,
follow up item (1) in https://github.com/pytorch/pytorch/pull/92641

Previously people will configure the observer/fq class in the following manner:

node.meta["target_dtype_info"] = {
  "input_act_obs_or_fq_ctr": obs1,
  "weight_obs_or_fq_ctr": obs2,
  "bias_obs_or_fq_ctr": obs3,
  "output_obs_or_fq_ctr": obs4,
}

This has hardcoded "weight" and "bias" and we are storing the index for weight and bias in BackendConfig,

this works for existing use cases but it's not general enough, since people may need to observe non-weight Tensors, and it's possible to use different
types of observers for each of them, and the input args may have more complicated structure.
```
node.meta["target_dtype_info"] = {
  "input_obs_or_fq_ctrs": (obs1, obs2, obs3),
  "output_obs_or_fq_ctr": obs4,
}

node.meta["target_dtype_info"] = {
  # this matches node.args
  "input_obs_or_fq_ctrs": (obs1, (obs2, obs3),)
  "output_obs_or_fq_ctr": obs4,
}
```

Next: do similar changes for output

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Refactor node.meta["target_dtype_info"] to allow generalized observer/fq configuration for input"


Summary:
This is a follow up PR to generalize the way of how users could configure what observer/fq class they want to use for each input of the node,
follow up item (1) in https://github.com/pytorch/pytorch/pull/92641

Previously people will configure the observer/fq class in the following manner:

node.meta["target_dtype_info"] = {
  "input_act_obs_or_fq_ctr": obs1,
  "weight_obs_or_fq_ctr": obs2,
  "bias_obs_or_fq_ctr": obs3,
  "output_obs_or_fq_ctr": obs4,
}

This has hardcoded "weight" and "bias" and we are storing the index for weight and bias in BackendConfig,

this works for existing use cases but it's not general enough, since people may need to observe non-weight Tensors, and it's possible to use different
types of observers for each of them, and the input args may have more complicated structure.
```
node.meta["target_dtype_info"] = {
  "input_obs_or_fq_ctrs": (obs1, obs2, obs3),
  "output_obs_or_fq_ctr": obs4,
}

node.meta["target_dtype_info"] = {
  # this matches node.args
  "input_obs_or_fq_ctrs": (obs1, (obs2, obs3),)
  "output_obs_or_fq_ctr": obs4,
}
```

Next: do similar changes for output

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44389339).

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -116,22 +117,43 @@
 # list of dtypes to not add observers to
 _DO_NOT_OBS_DTYPE_LIST = [int, float, torch.bool, None]
 
+_OBS_OR_FQ_TYPE = Union[ObserverBase, FakeQuantizeBase, Any]
```
Why Any here?

jerryzh168 replied:
Any is for things like MiinMaxObserver.with_args(...), maybe we can improve this part as well in the new api

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -116,22 +117,43 @@
+#    "input_obs_or_fq_ctrs": (obs1, obs2),
+#    "output_obs_or_fq_ctrs": (obs3, obs4)
+# }
+_OBS_OR_FQ_ANNO_TYPE = Union[_OBS_OR_FQ_TYPE, Tuple[_OBS_OR_FQ_TYPE, ...]]
```
Tuple can contain other tuples, right? SO I would imagine this would have to be
```
_OBS_OR_FQ_ANNO_TYPE = Union[_OBS_OR_FQ_TYPE, Tuple[_OBS_OR_FQ_ANNO_TYPE]]
```

jerryzh168 replied:
yeah, but does it support recursive definition?

kimishpatel reviewed (approved):
Looks good. Left some minor comments

jerryzh168 committed:
Update on "[quant][pt2e] Refactor node.meta["target_dtype_info"] to allow generalized observer/fq configuration for input"


Summary:
This is a follow up PR to generalize the way of how users could configure what observer/fq class they want to use for each input of the node,
follow up item (1) in https://github.com/pytorch/pytorch/pull/92641

Previously people will configure the observer/fq class in the following manner:

node.meta["target_dtype_info"] = {
  "input_act_obs_or_fq_ctr": obs1,
  "weight_obs_or_fq_ctr": obs2,
  "bias_obs_or_fq_ctr": obs3,
  "output_obs_or_fq_ctr": obs4,
}

This has hardcoded "weight" and "bias" and we are storing the index for weight and bias in BackendConfig,

this works for existing use cases but it's not general enough, since people may need to observe non-weight Tensors, and it's possible to use different
types of observers for each of them, and the input args may have more complicated structure.
```
node.meta["target_dtype_info"] = {
  "input_obs_or_fq_ctrs": (obs1, obs2, obs3),
  "output_obs_or_fq_ctr": obs4,
}

node.meta["target_dtype_info"] = {
  # this matches node.args
  "input_obs_or_fq_ctrs": (obs1, (obs2, obs3),)
  "output_obs_or_fq_ctr": obs4,
}
```

Next: do similar changes for output

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D44389339](https://our.internmc.facebook.com/intern/diff/D44389339)

[ghstack-poisoned]

vkuzo commented:
```
node.meta["target_dtype_info"] = {
  "input_obs_or_fq_ctrs": (obs1, obs2, obs3),
  "output_obs_or_fq_ctr": obs4,
}
```

would this be a good time to switch to "qdqstub" or "stub" instead of "obs_or_fq"?  We chatted before how it's technical debt that obs/fq is used for a generic "information holder", if we are changing the API this would be a low cost time to also clean up that abstraction.

jerryzh168 commented:
> ```
> node.meta["target_dtype_info"] = {
>   "input_obs_or_fq_ctrs": (obs1, obs2, obs3),
>   "output_obs_or_fq_ctr": obs4,
> }
> ```
> 
> would this be a good time to switch to "qdqstub" or "stub" instead of "obs_or_fq"? We chatted before how it's technical debt that obs/fq is used for a generic "information holder", if we are changing the API this would be a low cost time to also clean up that abstraction.

I thought qdqstub is going to be used in the flow as a placeholder for q/dq ops, but for information holder we still want to use observer/fqs?

vkuzo commented:
> > ```
> > node.meta["target_dtype_info"] = {
> >   "input_obs_or_fq_ctrs": (obs1, obs2, obs3),
> >   "output_obs_or_fq_ctr": obs4,
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > would this be a good time to switch to "qdqstub" or "stub" instead of "obs_or_fq"? We chatted before how it's technical debt that obs/fq is used for a generic "information holder", if we are changing the API this would be a low cost time to also clean up that abstraction.
> 
> I thought qdqstub is going to be used in the flow as a placeholder for q/dq ops, but for configuration we still want to use observer/fqs?

makes sense

jerryzh168 commented:
@pytorchbot merge

jerryzh168 committed:
Update on "[quant][pt2e] Refactor node.meta["target_dtype_info"] to allow generalized observer/fq configuration for input"


Summary:
This is a follow up PR to generalize the way of how users could configure what observer/fq class they want to use for each input of the node,
follow up item (1) in https://github.com/pytorch/pytorch/pull/92641

Previously people will configure the observer/fq class in the following manner:

node.meta["target_dtype_info"] = {
  "input_act_obs_or_fq_ctr": obs1,
  "weight_obs_or_fq_ctr": obs2,
  "bias_obs_or_fq_ctr": obs3,
  "output_obs_or_fq_ctr": obs4,
}

This has hardcoded "weight" and "bias" and we are storing the index for weight and bias in BackendConfig,

this works for existing use cases but it's not general enough, since people may need to observe non-weight Tensors, and it's possible to use different
types of observers for each of them, and the input args may have more complicated structure.
```
node.meta["target_dtype_info"] = {
  "input_obs_or_fq_ctrs": (obs1, obs2, obs3),
  "output_obs_or_fq_ctr": obs4,
}

node.meta["target_dtype_info"] = {
  # this matches node.args
  "input_obs_or_fq_ctrs": (obs1, (obs2, obs3),)
  "output_obs_or_fq_ctr": obs4,
}
```

Next: do similar changes for output

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D44389339](https://our.internmc.facebook.com/intern/diff/D44389339)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44389339).

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44389339).

jerryzh168 commented:
deprioritized for now, we can redo this after we get more user data
